firewood/src/merkle/node.rs

```diff
@@ -459,6 +463,203 @@ impl Storable for Node {
     }
 }
 
+pub struct EncodedNode<T> {
+    pub(crate) node: EncodedNodeType,
+    pub(crate) phantom: PhantomData<T>,
+}
+
+impl<T> EncodedNode<T> {
+    pub fn new(node: EncodedNodeType) -> Self {
+        Self {
+            node,
+            phantom: PhantomData,
+        }
+    }
+}
+pub enum EncodedNodeType {
+    Leaf(LeafNode),
+    Branch {
+        children: Box<[Option<Vec<u8>>; BranchNode::MAX_CHILDREN]>,
+        value: Option<Data>,
+    },
+}
+
+// Note that the serializer passed in should always be the same type as T in EncodedNode<T>.
+impl Serialize for EncodedNode<Bincode> {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+
+        match &self.node {
+            EncodedNodeType::Leaf(n) => {
+                let list = [
+                    Encoded::Raw(from_nibbles(&n.path.encode(true)).collect()),
+                    Encoded::Raw(n.data.to_vec()),
+                ];
+                let mut seq = serializer.serialize_seq(Some(list.len()))?;
+                for e in list {
+                    seq.serialize_element(&e)?;
+                }
+                seq.end()
+            }
+            EncodedNodeType::Branch { children, value } => {
+                let mut list = <[Encoded<Vec<u8>>; BranchNode::MAX_CHILDREN + 1]>::default();
+
+                for (i, c) in children.iter().enumerate() {
+                    if let Some(c) = c {
+                        if c.len() >= TRIE_HASH_LEN {
+                            let serialized_hash =
+                                Bincode::serialize(&Keccak256::digest(c).to_vec())
+                                    .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                            list[i] = Encoded::Data(serialized_hash);
+                        } else {
+                            list[i] = Encoded::Raw(c.to_vec());
+                        }
```

Comment:
You just need as `as_ref`:
```suggestion
                for (i, c) in children
                    .iter()
                    .enumerate()
                    .filter_map(|(i, c)| c.as_ref().map(|c| (i, c)))
                {
                    if c.len() >= TRIE_HASH_LEN {
                        let serialized_hash = Bincode::serialize(&Keccak256::digest(c).to_vec())
                            .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
                        list[i] = Encoded::Data(serialized_hash);
                    } else {
                        list[i] = Encoded::Raw(c.to_vec());
```